src/MSBuild/JsonOutputFormatter.cs

```diff
@@ -0,0 +1,151 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.CommandLine
+{
+    internal class JsonOutputFormatter
+    {
+        private static readonly JsonSerializerOptions Options = new() { AllowTrailingCommas = false, WriteIndented = true };
+        private readonly JsonNode _topLevelNode = new JsonObject();
+
+        public override string ToString()
+        {
+            return _topLevelNode.ToJsonString(Options);
+        }
+
+        internal void AddPropertiesInJsonFormat(string[] propertyNames, Func<string, string> getProperty)
```

Comment:
I see this as adding to the _topLevelNode object. It's a list of all the properties/items/results we've encountered so far, and each time one of these methods is called, more information is added to it.

Reply:
That's what the name suggests, yes. But when called multiple times, properties are not added. Instead the last call wins and overwrites everything added previously (line 41 `_topLevelNode["Properties"] = propertiesNode;`).

Just a nit really, feel free to ignore.

Reply:
I guess that scenario isn't too important to me because I know that it should not be called more than once. Would you prefer if I add a VerifyThrow that the "Properties" node doesn't exist yet?

Reply:
That would work, yes, thank you.